[OpenACC] Switch Clang to use the Flang 'appertainment' rules for cla…

erichkeane opened the pull request:
…uses

The Flang implemenation of OpenACC uses a .td file in the llvm/Frontend directory to determine appertainment in 4 categories:

-Required: If this list has items in it, the directive requires at least 1 of these be present.

-AllowedExclusive: Items on this list are all allowed, but only 1 from the list may be here (That is, they are exclusive of eachother).

-AllowedOnce: Items on this list are all allowed, but may not be duplicated.

Allowed: Items on this list are allowed.  Note th at the actual list of 'allowed' is all 4 of these lists together.

This is a draft patch to swtich Clang over to use those tables. Surgery to get this to happen in Clang Sema was somewhat reasonable. However, some gaps in the implementations are obvious, the existing clang implementation disagrees with the Flang interpretation of it. SO, we're keeping a task list here based on what gets discovered.

Changes to Clang:
- [x] Switch 'directive-kind' enum conversions to use tablegen See ff1a7bddd9435b6ae2890c07eae60bb07898bbf5
- [x] Switch 'clause-kind' enum conversions to use tablegen See ff1a7bddd9435b6ae2890c07eae60bb07898bbf5
- [x] Investigate 'parse' test differences to see if any new disagreements arise.
- [x] Clang/Flang disagree as to whether 'collapse' can be multiple times on a loop.  Further research showed no prose to limit this, and the comment on the clang implementation said "no good reason to allow", so no standards justification.
- [x] Clang/Flang disagree whether 'num_gangs' can appear >1 on a compute/combined construct. This ended up being an unjustified restriction.
- [x] Clang/Flang disagree as to the list of required clauses on a 'set' construct. My research shows that Clang mistakenly included 'if' in the list, and that it should be just 'default_async', 'device_num', and 'device_type'.
- [x] Order of 'at least one of' diagnostic has changed. Tests were updated.
- [x] Ensure we are properly 'de-aliasing' clause names in appertainment checks?
- [x] What is 'shortloop'?  'shortloop' seems to be an old non-standard extension that isn't supported by flang, but is parsed for backward compat reasons.  Clang won't parse, but we at least have a spot for it in the clause list.
- [x] Implemented proposed change for 'routine' gang/worker/vector/seq.  (see issue 539)
- [x] Implement init/shutdown can only have 1 'if' (see issue 540)
- [x] Clang/Flang disagree as to whether 'tile' is permitted more than once on a 'loop' or combined constructs (Flang prohibits >1).  I see no justification for this in the standard.  EDIT: I found a comment in clang that I did this to make SOMETHING around duplicate checks easier.  Discussion showed we should actually have a better behavior around 'device_type' and duplicates, so I've since implemented that.
- [x] Clang/Flang disagree whether 'gang', 'worker', or 'vector' may appear on the same construct as a 'seq' on a 'loop' or 'combined'. There is prose for this in 2022: (a gang, worker, or vector clause may not appear if a 'seq' clause appears). EDIT: These don't actually disagree, but aren't in the .td file, so I restored the existing code to do this.
- [x]  Clang/Flang disagree on whether 'bind' can appear >1 on a 'routine'.  I believe line 3096 (A bind clause may not bind to a routine name that has a visible bind clause) makes this limitation (Flang permits >1 bind). we discussed and decided this should have the same rules as worker/vector/etc, except without the 'exactly 1 of' rule (so no dupes in individual sections).
- [x] Clang/Flang disagree on whether 'init'/'shutdown' can have multiple 'device_num' clauses.  I believe there is no supporting prose for this limitation.,  We decided that `device_num` should only happen 1x.
- [x] Clang/Flang disagree whether 'num_gangs' can appear >1 on a 'kernels'  construct. Line 1173 (On a kernels construct, the num_gangs clause must have a single argument) justifies limiting on a per-arguement basis, but doesn't do so for multiple num_gangs clauses. WE decided to do this with the '1-per-device-type' region for num_gangs, num_workers, and vector_length, see openacc bug here: https://github.com/OpenACC/openacc-spec/issues/541

Changes to Flang:
- [x] Clang/Flang disgree on whether 'atomic' can take an 'if' clause. This was added in OpenACC3.3_Next See #135451
- [x] Clang/Flang disagree on whether 'finalize' can be allowed >1 times on a 'exit_data' construct.  see #135415.
- [x] Clang/Flang disagree whether 'if_present' should be allowed >1 times on a 'host_data'/'update' construct.  see #135422
- [x] Clang/Flang disagree on whether 'init'/'shutdown' can have multiple 'device_type' clauses. I believe there is no supporting prose for this limitation.
- [ ] SEE change for num_gangs/etc above.


Changes that need discussion/research:

erichkeane committed:
[OpenACC] Switch Clang to use the Flang 'appertainment' rules for clauses

The Flang implemenation of OpenACC uses a .td file in the llvm/Frontend
directory to determine appertainment in 4 categories:

-Required: If this list has items in it, the directive requires at least
1 of these be present.

-AllowedExclusive: Items on this list are all allowed, but only 1 from
the list may be here (That is, they are exclusive of eachother).

-AllowedOnce: Items on this list are all allowed, but may not be
duplicated.

Allowed: Items on this list are allowed.  Note th at the actual list of
'allowed' is all 4 of these lists together.

This is a draft patch to swtich Clang over to use those tables. Surgery
to get this to happen in Clang Sema was somewhat reasonable. However,
some gaps in the implementations are obvious, the existing clang
implementation disagrees with the Flang interpretation of it. SO, we're
keeping a task list here based on what gets discovered.

Changes to Clang:
- [ ] Switch 'directive-kind' enum conversions to use tablegen
- [ ] Switch 'clause-kind' enum conversions to use tablegen
- [x] Investigate 'parse' test differences to see if any new
disagreements arise.
- [x] Clang/Flang disagree as to whether 'collapse' can be multiple
times on a loop.  Further research showed no prose to limit this, and
the comment on the clang implementation said "no good reason to allow",
so no standards justification.
- [x] Clang/Flang disagree whether 'num_gangs' can appear >1 on a
compute/combined construct. This ended up being an unjustified
restriction.
- [x] Clang/Flang disagree as to the list of required clauses on a 'set'
construct. My research shows that Clang mistakenly included 'if' in the
list, and that it should be just 'default_async', 'device_num', and
'device_type'.
- [x] Order of 'at least one of' diagnostic has changed. Tests were
updated.

Changes to Flang:
- [ ] Clang/Flang disgree on whether 'atomic' can take an 'if' clause.
This was added in OpenACC3.3_Next

Changes that need discussion/research:
- [ ] Clang/Flang disagree on whether 'init'/'shutdown' can have
multiple 'device_type' clauses. I believe there is no supporting prose
for this limitation.
- [ ] Clang/Flang disagree on whether 'init'/'shutdown' can have
multiple 'device_num' clauses.  I believe there is no supporting prose
for this limitation.,
- [ ] Clang/Flang disagree on whether 'init'/'shutdown' can have
multiple 'if' clauses. I believe there is no supporting prose for this
limitation.
- [ ]  Clang/Flang disagree on whether 'bind' can appear >1 on a
'routine'.  I believe line 3096 (A bind clause may not bind to a routine
    name that has a visible bind clause) makes this limitation.
- [ ] Clang/Flang disagree on whether 'gang', 'worker', 'vector', 'seq'
may appear multiple times on a 'routine'.  Line 3080(Exactly one of the
    gang, worker, vector, or seq clauses must appear).  It is the flang
opinion that it means once-per-'device_type' (and once before each
    device_type).
- [ ] Clang/Flang disagree as to whether 'tile' is permitted more than
once on a 'loop' or combined constructs (Flang prohibits >1).  I see no
justification for this in the standard.
- [ ] Clang/Flang disagree whether 'loop' or combined constructs can
have 'seq', 'independent', or 'auto' with a 'device_type' clause. I see
no prose to suggest this limitation.
- [ ] Clang/Flang disagree whether 'gang', 'worker', or 'vector' may
appear on the same construct as a 'seq' on a 'loop' or 'combined'. There
is prose for this in 2022: (a gang, worker, or vector clause may not
    appear if a 'seq' clause appears).
- [ ] Clang/Flang disagree on clause list that may appear with a
'auto'/'independent', or 'seq' clause. for a 'loop' or 'combined'.
Flang seems to be rejecting everything, which doesn't seem correct to
me, as I can find no supporting prose. This MIGHT be a bug/difference in
implementation?
- [ ] Clang/Flang disagree whether 'num_gangs' can appear >1 on a
'kernels'  construct. Line 1173 (On a kernels construct, the num_gangs
clause must have a single argument) justifies limiting on a per-clause
basis, but doesn't do so for multiple num_gangs. ALSO: IF we decide such
a restriction means this, it needs to be applied to `kernels loop` as
well.
- [ ] Clang/Flang disagree whether 'finalize' and 'if_present' should be
allowed >1 times on a 'exit_data' construct.  I see no prose to support
this?
- [ ] Clang/Flang disagree whether 'seq' and 'self' can appear on the
same serial-loop construct.  I see no prose to support this?

erichkeane commented:
@razvanlupusoru @clementval @jefflarkin : This is the place I discussed/promised.

Note I used 'flang' for a stand-in as "after using the Acc.td file in clang", so some of those might not be perfectly accurate.  BUT are ones that need discussion in some way.

We can discuss the changes to OpenACC Spec via email/discussion, as it probably keeps them best organized (and perhaps 1x at a time?).  As you can see, there is plenty to do/discuss :)

erichkeane commented:
Note: No real reason to review this quite yet?  I'm still going to do a bunch of work on tablegen to no  longer use the FLANG generated things by generating clang ones that are similar, so a lot of the Appertainment.cpp file is going to change.

erichkeane committed:
Fix bug with 'exclusive clauses'

erichkeane commented:
Note: I discovered a bug in my 'exclusive' check that eliminated 2 items from the above list!

erichkeane committed:
Starting effort to make tablegen specifically for clang.x

clementval commented:
> Clang/Flang disagree whether 'seq' and 'self' can appear on the same serial-loop construct. I see no prose to support this?

`seq` and `self` are not in the same set for `serial loop` and seems to be accepted. https://godbolt.org/z/dzdoE3xKr

erichkeane committed:
Modify tablegen to emit something clang can just consume.

In an attempt to reduce the amount of hackery that clang had to do, I've
added some tablegen stuff.  There is some things that are perhaps not
well isolated, but hopefully this'll improve the codebase at least a
little.

erichkeane committed:
Clang-format for tablegen changes

erichkeane commented:
> > Clang/Flang disagree whether 'seq' and 'self' can appear on the same serial-loop construct. I see no prose to support this?
> 
> `seq` and `self` are not in the same set for `serial loop` and seems to be accepted. https://godbolt.org/z/dzdoE3xKr

This one seems to be a mistake with how I was doing 'exclusive'.  I struck out some of them, so I'll strike out this one as well.

erichkeane committed:
Merge branch 'main' into ACCtdAppertainment

erichkeane commented:
Note tablegen changes undid the 'dealiasing' that we were doing in the manual conversion between types, so many more tests will fail.  Working now to add dealiasing to these functions.

erichkeane committed:
Dealias clauses correctly, which should eliminate failures after tablegen.  Also, sort the clause kind list so that diagnostics are alphabetical

erichkeane committed:
Merge branch 'main' into ACCtdAppertainment

erichkeane committed:
Clang-format

erichkeane committed:
Correct implementation of 'device_type' on routine

After looking into it further, the rule for the 'worker', 'gang',
'vector', and 'seq' on a 'routine' construct are suppsoed to be more
complicated.

The area BEFORE the first device_type are 'global' and apply to all
following device_types.

Then, each device_type region can name their own.

The rule is it is exclusive on a per-device_type basis for these.

erichkeane committed:
Merge remote-tracking branch 'origin/main' into ACCtdAppertainment

erichkeane commented:
> `if_present` #135422 merged `finalize` #135415 merged

Picked up in the latest merge, just committed.

erichkeane commented:
That latest patch that implements the 'routine' worker/seq/vector/gang restrictions reduces us to 13 test failures in clang!  

This is my analysis of them:
```
  
FLANG: `atomic-if'  (Valentin has  PR)
    Clang :: SemaOpenACC/atomic-construct.cpp
    Clang :: ParserOpenACC/parse-clauses.c
    Clang :: AST/ast-print-openacc-atomic-construct.cpp
    Clang :: ParserOpenACC/parse-clauses.cpp
    Clang :: SemaOpenACC/atomic-construct-ast.cpp
  
FLANG: Doesn't prohibit gang/worker/vector/seq' on 'loop'/combined correctly (UNK way forward right now)
    Clang :: SemaOpenACC/loop-construct-auto_seq_independent-clauses.c
    Clang :: SemaOpenACC/combined-construct-auto_seq_independent-clauses.c
  
FLANG: 'tile' prohibited >1x (Standard doesn't support, UNK way forward right now)
    Clang :: SemaOpenACC/loop-construct-tile-clause.cpp
    Clang :: SemaOpenACC/combined-construct-tile-clause.cpp
  
Disagreements on 'num_gangs' appearing >1x on kernels/kernels-loop (justification for kernels-loop should apply to kernels as well?) (UNK way forward right now)
    Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
  
FLANG: prohibits device_num/if >1x  ('if' probably should be prohibited, i'll write a patch to do so) (device_num UNK way forward RN).
  Clang :: SemaOpenACC/init-construct.cpp
  Clang :: SemaOpenACC/shutdown-construct.cpp
  
Disagreement on 'bind' on a 'routine'. (UNK way forward right now)
   Clang :: SemaOpenACC/routine-construct-clauses.cpp
```

erichkeane committed:
Update 'init' and 'shutdown' tests to accept that multiples of 'if' are ill-formed

erichkeane committed:
Allow multiple tile clauses, fix multiple 'collapse' checking

There is no restriction on multiple tile clauses.  There is perhaps some
goofy behavioral decisions to make during codegen to decide which one
wins, but here it isn't too bad.

Multiple 'collapse' clauses have a similar problem, but with more
difficult checking (since we have to take the instantiation status into
    consideration).

erichkeane committed:
correctly diagnose duplicate 'tile' and 'collapse' clauses

After discussion, it is clear that we should do some funny-business
around 'device_type' on 'tile' and 'collapse', so this patch implements
that checking.

erichkeane commented:
> FLANG: `atomic-if'  (Valentin has  PR)
>     Clang :: SemaOpenACC/atomic-construct.cpp
>     Clang :: ParserOpenACC/parse-clauses.c
>     Clang :: AST/ast-print-openacc-atomic-construct.cpp
>     Clang :: ParserOpenACC/parse-clauses.cpp
>     Clang :: SemaOpenACC/atomic-construct-ast.cpp
>   
> FLANG: Doesn't prohibit gang/worker/vector/seq' on 'loop'/combined correctly (UNK way forward right now)
>     Clang :: SemaOpenACC/loop-construct-auto_seq_independent-clauses.c
>     Clang :: SemaOpenACC/combined-construct-auto_seq_independent-clauses.c
>   
> FLANG: 'tile' prohibited >1x (Standard doesn't support, UNK way forward right now)
>     Clang :: SemaOpenACC/loop-construct-tile-clause.cpp
>     Clang :: SemaOpenACC/combined-construct-tile-clause.cpp
>   
> Disagreements on 'num_gangs' appearing >1x on kernels/kernels-loop (justification for kernels-loop should apply to kernels as well?) (UNK way forward right now)
>     Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
>   
> FLANG: prohibits device_num >1x  (UNK way forward RN).
>   Clang :: SemaOpenACC/init-construct.cpp
>   Clang :: SemaOpenACC/shutdown-construct.cpp
>   
> Disagreement on 'bind' on a 'routine'. (UNK way forward right now)
>    Clang :: SemaOpenACC/routine-construct-clauses.cpp

```
FLANG: `atomic-if'  (Valentin has  PR)
    Clang :: SemaOpenACC/atomic-construct.cpp
    Clang :: ParserOpenACC/parse-clauses.c
    Clang :: AST/ast-print-openacc-atomic-construct.cpp
    Clang :: ParserOpenACC/parse-clauses.cpp
    Clang :: SemaOpenACC/atomic-construct-ast.cpp
  
FLANG: Doesn't prohibit gang/worker/vector/seq' on 'loop'/combined correctly (UNK way forward right now)
    Clang :: SemaOpenACC/loop-construct-auto_seq_independent-clauses.c
    Clang :: SemaOpenACC/combined-construct-auto_seq_independent-clauses.c
  
Disagreements on 'num_gangs' appearing >1x on kernels/kernels-loop (justification for kernels-loop should apply to kernels as well?) (UNK way forward right now)
    Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
  
FLANG: prohibits device_num >1x  (UNK way forward RN).
  Clang :: SemaOpenACC/init-construct.cpp
  Clang :: SemaOpenACC/shutdown-construct.cpp
  
Disagreement on 'bind' on a 'routine'. (UNK way forward right now)
   Clang :: SemaOpenACC/routine-construct-clauses.cpp
```

Updated list!

erichkeane committed:
Merge branch 'main' into ACCtdAppertainment

erichkeane commented:
atomic-if merged, which leaves us at:
```
Failed Tests (6):
  Clang :: SemaOpenACC/combined-construct-auto_seq_independent-clauses.c
  Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
  Clang :: SemaOpenACC/init-construct.cpp
  Clang :: SemaOpenACC/loop-construct-auto_seq_independent-clauses.c
  Clang :: SemaOpenACC/routine-construct-clauses.cpp
  Clang :: SemaOpenACC/shutdown-construct.cpp
```

Same categorization as above.

erichkeane committed:
Restore the seq/gang/worker/vector combination rules on loops

erichkeane commented:
We're in agreement on all of the 'standards discussion needed' part, and we've filed bugs for them, so i've removed them.  So now we're down to 3 differences, which I will spawn email threads for between us and the OpenACC chair.

erichkeane committed:
Implement 'bind' restrictions on routine like we previously did for
gang/worker/vector

Discussed separately, this seems like the right way forward.

erichkeane commented:
> > Updated list:
> > ```
> > Disagreements on 'num_gangs' appearing >1x on kernels/kernels-loop (justification for kernels-loop should apply to kernels as well?) (UNK way forward right now)
> >     Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
> >   
> > FLANG: prohibits device_num >1x  (UNK way forward RN).
> >   Clang :: SemaOpenACC/init-construct.cpp
> >   Clang :: SemaOpenACC/shutdown-construct.cpp
> >   
> > Disagreement on 'bind' on a 'routine'. (UNK way forward right now)
> >    Clang :: SemaOpenACC/routine-construct-clauses.cpp
> > ```
> 
> In discussion, we decided that `bind` probably should follow the same rules as "Implemented proposed change for 'routine' gang/worker/vector/seq. (see issue 539)". So I'll follow-up on that after this.



We discussed/solved 'bind', so:
```
 Disagreements on 'num_gangs' appearing >1x on kernels/kernels-loop (justification for kernels-loop should apply to kernels as well?) (UNK way forward right now)
     Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
   
 FLANG: prohibits device_num >1x  (UNK way forward RN).
   Clang :: SemaOpenACC/init-construct.cpp
   Clang :: SemaOpenACC/shutdown-construct.cpp
```

erichkeane committed:
Merge branch 'main' into ACCtdAppertainment

erichkeane committed:
Merge branch 'main' into ACCtdAppertainment

erichkeane committed:
Update init/shutdown tests now that we agree it should only have device_num 1x

erichkeane commented:
> > > Updated list:
> > > ```
> > > Disagreements on 'num_gangs' appearing >1x on kernels/kernels-loop (justification for kernels-loop should apply to kernels as well?) (UNK way forward right now)
> > >     Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
> > >   
> > > FLANG: prohibits device_num >1x  (UNK way forward RN).
> > >   Clang :: SemaOpenACC/init-construct.cpp
> > >   Clang :: SemaOpenACC/shutdown-construct.cpp
> > >   
> > > Disagreement on 'bind' on a 'routine'. (UNK way forward right now)
> > >    Clang :: SemaOpenACC/routine-construct-clauses.cpp
> > > ```
> > 
> > 
> > In discussion, we decided that `bind` probably should follow the same rules as "Implemented proposed change for 'routine' gang/worker/vector/seq. (see issue 539)". So I'll follow-up on that after this.
> 
> We discussed/solved 'bind', so:
> 
> ```
>  Disagreements on 'num_gangs' appearing >1x on kernels/kernels-loop (justification for kernels-loop should apply to kernels as well?) (UNK way forward right now)
>      Clang :: SemaOpenACC/compute-construct-num_gangs-clause.c
>    
>  FLANG: prohibits device_num >1x  (UNK way forward RN).
>    Clang :: SemaOpenACC/init-construct.cpp
>    Clang :: SemaOpenACC/shutdown-construct.cpp
> ```

Implemented our agreement on the `device_num` on init/shutdown, so down to only the num_gangs/num_workers/vector_length restriction we proposed, which requires that the Appertainment files (and thus likely the Fortran impl) change.

@clementval

erichkeane committed:
Implement 'new' num_gangs/worker/vector_length rules around device_type, lots of test updates, fixed up the diagnostics-sema-kinds

erichkeane committed:
Remove unneeded forward decl

erichkeane commented:
@clementval @razvanlupusoru 

This should now be ready, though it probably has to be merged with the flang changes after `num_gangs`/`num_workers`/`vector_length` changes are made.

MOST of this ended up being test changes as we went along, But the tablegen/ACC.td changes/etc are mostly what I want you guys to look at.  SemaOpenACCAppertainment.cpp is where all of the changes related to using ACC.td are located.

clementval commented:
> @clementval @razvanlupusoru
> 
> This should now be ready, though it probably has to be merged with the flang changes after `num_gangs`/`num_workers`/`vector_length` changes are made.

Working on that right now. 

> MOST of this ended up being test changes as we went along, But the tablegen/ACC.td changes/etc are mostly what I want you guys to look at. SemaOpenACCAppertainment.cpp is where all of the changes related to using ACC.td are located.

Will check. Thanks for all the updates and making this consistent

clementval reviewed (approved):
LGTM. I didn't go deep into the clang part since I don't have much knowledge there. There rest looks good. I'll merge https://github.com/llvm/llvm-project/pull/136341 as soon as the pre-commit ci is done.

erichkeane committed:
Fix space removed during directive-tablegen

erichkeane committed:
Merge branch 'main' into ACCtdAppertainment

clementval commented:
You might be missing a dependency on the TableGen target so the file is generated.

erichkeane commented:
> You might be missing a dependency on the TableGen target so the file is generated.

Thanks!  I THINK I fixed it here: https://github.com/llvm/llvm-project/commit/eec1af21c0828544fc49ee1a982a9cdef455b92a
